private/bufpkg/bufmodule/targeting_module.go

```diff
@@ -92,53 +106,116 @@ func (m *targetingModule) TargetFileInfos(ctx context.Context) (fileInfos []bufm
+			excludeMatchingPathMap := normalpath.MapAllEqualOrContainingPathMap(
+				excludePathMap,
 				path,
 				normalpath.Relative,
 			)
-			if len(fileMatchingPathMap) > 0 {
-				// we had a match, this means that some path in potentialDirPaths matched
+			for excludeMatchingPath := range excludeMatchingPathMap {
+				if _, ok := matchingExcludePaths[excludeMatchingPath]; !ok {
+					matchingExcludePaths[excludeMatchingPath] = struct{}{}
+				}
+			}
+			var fileMatchingPathMap map[string]struct{}
+			if !m.allowAllTargetPaths {
+				// get the paths in potentialDirPathMap that match this path
+				fileMatchingPathMap = normalpath.MapAllEqualOrContainingPathMap(
+					potentialDirPathMap,
+					path,
+					normalpath.Relative,
+				)
+			}
+			// 1. If we find a match for both target and exclude paths, we need to resolve the union
+			// of the paths.
+			// 2. If we find it only for exclude paths, then we exclude.
+			// 3. If we do not find it for our target path and allowAllTargetPaths=false, we exclude.
+			// 4. If allowAllTargetPaths=false but we find it in our target paths, we need to add
+			// the file to `matchingPotentialDirPathMap` to check against the `pathsAllowNotExistOnWalk`
+			// condition.
+			// 5. In the remaining case, we add the file if `len(fileMatchingPathMap) > 0 || allowAllTargetPaths == true`.
+			if len(excludeMatchingPathMap) > 0 && len(fileMatchingPathMap) > 0 {
+				// We need to merge the results for exclude and target paths based on the most specific
+				// configuration.
+				// Case 1 - target path is more specific than exclude, we include:
+				//   --path a/b.proto --exclude a
+				// Case 2 - exclude path is more specific than target, we exclude:
+				//   --path a --exclude a/b.proto
+				// Case 3 - they are equal, we error:
+				//   --path a --exclude a
+				for fileMatchingPath := range fileMatchingPathMap {
+					for excludeMatchingPath := range excludeMatchingPathMap {
+						if fileMatchingPath == excludeMatchingPath {
+							return fmt.Errorf("cannot set the same path for both --path and --exclude flags: %s", excludeMatchingPath)
+						}
+						// Since we already checked for the equal condition, we know that this must be contain.
+						// In this case
+						if normalpath.EqualsOrContainsPath(fileMatchingPath, excludeMatchingPath, normalpath.Relative) {
+							return nil
+						}
+					}
+				}
+			}
+			// We have ruled out case 2 + 3 above, so now we are checking for case 1. In the situation
+			// where we are accepting all targets, then we can just move on. Otherwise, if no file
+			// matching any of the target paths were found, we continue the walk.
+			if !m.allowAllTargetPaths && len(fileMatchingPathMap) == 0 {
+				return nil
+			}
+			// We have checked all conditions where a matching file was found, so we need to handle
+			// if an exclude was found and we are allowing all targets.
+			if m.allowAllTargetPaths && len(excludeMatchingPathMap) > 0 {
+				return nil
+			}
```

Comment:
I'd like to see if we can simplify these cases regardless of the `m.allowAllTargetPaths` flag (re: the comment above).